File: Aulas/Outros/ASP.NET/Aula 02/ApiCatalogo/ApiCatalogo/Repositories/Repository.cs
using System.Linq.Expressions;
using ApiCatalogo.Context;
using Microsoft.EntityFrameworkCore;

namespace ApiCatalogo.Repositories;

public class Repository<T> : IRepository<T> where T : class
{
   protected readonly private AppDbContext _context;
   
   public Repository(AppDbContext context)
   {
      _context = context;
   }
   
   public async Task<IEnumerable<T>> GetAllAsync()
   {
      return await _context.Set<T>().AsNoTracking().ToListAsync();
   }

   public async Task<T?> GetAsync(Expression<Func<T, bool>> predicate)
   {
      return await _context.Set<T>().FirstOrDefaultAsync(predicate);
   }

   public T Create(T entity)
   {
      _context.Set<T>().Add(entity);
      //_context.SaveChanges();
      
      return entity;
   }

   public T Update(T entity)
   {
      _context.Set<T>().Update(entity);
     //_context.SaveChanges();
      return entity;
   }

   public T Delete(T entity)
   {
      _context.Set<T>().Remove(entity);
      //_context.SaveChanges();
      return entity;
   }
}

File: Aulas/Programação Script/2025-1/Aula 05/repositories/Repository.js
const IRepository = require('./IRepository');

class Repository extends IRepository {
    constructor(model) {
        super();
        this.model = model;
    }

    async create(data) {
        return await this.model.create({data});
    }

    async findAll(include = undefined) {
        return await this.model.findMany({ include });
    }

    async findById(id, include = undefined) {
        return await this.model.findUnique({ where: { id }, include });
    }

    async update(id, data) {
        return await this.model.update({ where: { id }, data });
    }

    async delete(id){
        return await this.model.delete({ where: { id }});
    }

}